core/src/main/scala/kafka/log/Log.scala

```diff
@@ -243,6 +255,26 @@ class Log(val dir: File,
       recoverLog()
       // reset the index size of the currently active log segment to allow more entries
       activeSegment.index.resize(config.maxIndexSize)
+      activeSegment.timeIndex.resize(config.maxIndexSize)
+    }
+
+    /**
+     * We append a time index entry for non-active segment with an empty time index. This could happen during upgrade.
+     * We need to do this because searching by timestamp needs a time index entry.
+     */
+    val logSegmentsSeq = logSegments.toSeq
+    for (i <- 0 until logSegmentsSeq.size) {
+      // load the largest timestamp into memory.
+      logSegmentsSeq(i).loadLargestTimestamp()
+      // Insert the last entry to the time index for inactive segments.
+      if (logSegmentsSeq(i) != activeSegment) {
+        // It is possible that the time index was created for the recovered or swapped log segment file and is
+        // resized to 0. We need to resize it to hold at least one entry.
+        if (logSegmentsSeq(i).timeIndex.maxEntries == 0)
+          logSegmentsSeq(i).timeIndex.resize(config.maxIndexSize)
+        logSegmentsSeq(i).ensureNonEmptyTimeIndex()
```

Comment:
Hmm, for those segments w/o time index because they were created in 0.9, we will update the index with the last modified time. However, it seems that we didn't change maxTimestampSoFar in the log segment after that since  logSegmentsSeq(i).loadLargestTimestamp() is called before updating the index?